contracts/core/fee-calculators/PerformanceFeeCalculator.sol

```diff
@@ -0,0 +1,383 @@
+        address indexed rebalancingSetToken,
+        uint256 oldHighWatermark,
+        uint256 newHighWatermark,
+        uint256 inflationFee
+    );
+
+    /* ============ Structs ============ */
+    struct FeeState {
+        uint256 profitFeeFrequency;             // Time required between accruing profit fees
+        uint256 highWatermarkResetFrequency;    // Time required after last profit fee to reset high watermark 
+        uint256 profitFeePercentage;            // Percent of profits that accrue to manager
+        uint256 streamingFeePercentage;         // Percent of Set that accrues to manager each year
+        uint256 highWatermark;                  // Value of Set at last profit fee accrual
+        uint256 lastProfitFeeTimestamp;         // Timestamp last profit fee was accrued
+        uint256 lastStreamingFeeTimestamp;      // Timestamp last streaming fee was accrued
+    }
+
+    struct InitFeeParameters {
+        uint256 profitFeeFrequency;
+        uint256 highWatermarkResetFrequency;
+        uint256 profitFeePercentage;
+        uint256 streamingFeePercentage;        
+    }
+
+    /* ============ Constants ============ */
+    // 365.25 days used to represent the year
+    uint256 private constant ONE_YEAR_IN_SECONDS = 31557600;
+    uint256 private constant ONE_HUNDRED_PERCENT = 1e18;
+
+    /* ============ State Variables ============ */
+    ICore public core;
+    IOracleWhiteList public oracleWhiteList;
+    uint256 public maximumProfitFeePercentage;
+    uint256 public maximumStreamingFeePercentage;
+    mapping(address => FeeState) public feeState;
+
+    /* ============ Constructor ============ */
+
+    /**
+     * Constructor function for PerformanceFeeCalculator
+     *
+     * @param _core                                 Core instance
+     * @param _oracleWhiteList                      Oracle white list instance
+     * @param _maximumProfitFeePercentage           Maximum percent of profit fee scaled by 1e18
+     *                                              (e.g. 100% = 1e18 and 1% = 1e16)
+     * @param _maximumStreamingFeePercentage        Maximum percent of streaming fee scaled by 1e18
+     *                                              (e.g. 100% = 1e18 and 1% = 1e16)
+     */
+    constructor(
+        ICore _core,
+        IOracleWhiteList _oracleWhiteList,
+        uint256 _maximumProfitFeePercentage,
+        uint256 _maximumStreamingFeePercentage
+    )
+        public
+    {
+        core = _core;
+        oracleWhiteList = _oracleWhiteList;
+        maximumProfitFeePercentage = _maximumProfitFeePercentage;
+        maximumStreamingFeePercentage = _maximumStreamingFeePercentage;
+    }
+
+    /* ============ External Functions ============ */
+
+    /*
+     * Called by RebalancingSetToken, parses bytedata then assigns to correct FeeState struct. 
+     *
+     * @param  _feeCalculatorData       Bytestring encoding fee parameters for RebalancingSetToken
+     */
+    function initialize(
```

Comment:
Right now the creation flow is as follows:

Set gets created -> Factory calls initialize on the Set -> Set calls FeeCalculator.

This happens all in one transaction. I don't see how someone else would be able to front-run or anything. Either 1) the Set does not exist yet or 2) the caller is not the Set.